chanzuckerberg/czgenepi
# thuang-e2e-part-1-123908

### Description

This PR is part 1 of 3 for complete E2E test setup!

Part 2 is getting E2E to run per PR
Part 3 is getting E2E to run after each staging and prod deployment (which also runs `describeIfDeployed` tests)

Changes:
1. Set up E2E tests using Jest + [Playwright](https://playwright.dev/), which is a rewrite of Puppeteer from the original Google team that's at Microsoft now 🎭 
1. Add tests for the following features:
    1. Assert Samples Page and Trees Page render some expected elements
    1. Login works (which is implicit, since many tests require login first)
    1. Search works (we will search for the first public id, and assert the table only contains one row)
    2. Sort works (we sort by public id header, and assert the order is different from the default sort order (upload date))
    1. Test Nextstrain link is created (only run this test in deployed environment. E.g., staging and prod)
    1. Test TSV file link is generated when Samples Page's table is rendered
1. Update `npm run test` to `npm run tests`, since autocorrect always bugs me about using plural 
2. Add `npm run e2e` command to run against local FE server: `http://frontend.genepinet.local:8000`
3. Add `npm run e2e-staging-from-local` to run E2E tests against staging: `https://staging.genepi.czi.technology`. **THIS IS HELPFUL FOR ONCALL ENGINEERS TO SMOKE TEST STAGING!** After part 3 is merged, we won't need to do this anymore
4. Update unit tests to accommodate [`react-query`](https://react-query.tanstack.com/), which we use for fetching and managing data from API endpoints. This involves setting up [Mock Server Worker](https://mswjs.io/) to return mocked data

#### Issue
https://app.clubhouse.io/genepi/story/123908/cross-browser-testing-functionality

### Test plan

1. Pull the branch
2. Run `make local-sync && make local-init && make frontend-e2e`

All tests should pass!

PTAL thanks all 😄 🎉 !
